extra/mariadbfrm/mariadb_frm.cc

```diff
@@ -0,0 +1,531 @@
+  {
+    fprintf(stderr, "Error: Cannot extract database and table names from path\n");
+    goto cleanup;
+  }
+  
+  printf("DEBUG: Names extracted - db: %.*s, table: %.*s\n", 
+         (int)db_name.length, db_name.str, (int)table_name.length, table_name.str);
+  fflush(stdout);
+
+  printf("DEBUG: Allocating TABLE_SHARE manually\n");
+  fflush(stdout);
+  
+  share = (TABLE_SHARE*)my_malloc(PSI_NOT_INSTRUMENTED, sizeof(TABLE_SHARE), MYF(MY_ZEROFILL));
+  if (!share)
+  {
+    fprintf(stderr, "Error: Cannot allocate TABLE_SHARE\n");
+    goto cleanup;
+  }
+  
+  share->db.str = db_name.str;
+  share->db.length = db_name.length;
+  share->table_name.str = table_name.str;
+  share->table_name.length = table_name.length;
+  
+  init_alloc_root(PSI_NOT_INSTRUMENTED, &share->mem_root, 1024, 0, MYF(0));
+  
+  fake_thd->mem_root = &share->mem_root;
+  
+  printf("DEBUG: TABLE_SHARE allocated and initialized successfully\n");
+  fflush(stdout);
+  
+  printf("DEBUG: About to call init_from_binary_frm_image\n");
+  fflush(stdout);
+  
+  if (share->init_from_binary_frm_image((THD*)fake_thd, false, frm_data, frm_length))
+  {
+    fprintf(stderr, "Error: Cannot parse FRM file - init_from_binary_frm_image failed\n");
+    goto cleanup;
+  }
+  
+  printf("DEBUG: init_from_binary_frm_image completed successfully\n");
+  fflush(stdout);
+  
+
+  table= static_cast<TABLE *>(
+      my_malloc(PSI_NOT_INSTRUMENTED, sizeof(TABLE), MYF(MY_ZEROFILL)));
+  if (!table)
+  {
+    fprintf(stderr, "Error: Cannot allocate TABLE structure\n");
+    goto cleanup;
+  }
+  
+  table->s= share;
+  table->in_use= (THD*)fake_thd;
+  
+  table_list.table_name= Lex_ident_table(table_name);
+  table_list.db= Lex_ident_db(db_name);
+  table_list.alias= Lex_ident_table(table_name);
+  
+  table->init((THD*)fake_thd, &table_list);
+  
+  printf("Table: %.*s.%.*s\n", (int)db_name.length, db_name.str, (int)table_name.length, table_name.str);
+  printf("Fields: %d\n", share->fields);
+  printf("Keys: %d\n", share->keys);
+  
+  error= false;
+  
+cleanup:
+  if (share)
+  {
+    free_root(&share->mem_root, MYF(0));
+    my_free(share);
+  }
+  my_free(table);
+  my_free(frm_data);
+  my_free((void*)db_name.str);
+  my_free((void*)table_name.str);
+  return error;
+}
+
+/**
+  Main function
+*/
+int main(int argc, char **argv)
+{
+  printf("DEBUG: Starting frm_parser...\n");
+  fflush(stdout);
+  
+  MY_INIT(argv[0]);
+  printf("DEBUG: MY_INIT completed\n");
+  fflush(stdout);
+  
+  THD *fake_thd= NULL;
+  int exit_code= 0;
+
+  if (argc < 2)
+  {
+    fprintf(stderr, "Usage: %s <frm_file>\n", argv[0]);
+    return 1;
+  }
+
+  printf("DEBUG: Arguments validated, FRM file: %s\n", argv[1]);
+  fflush(stdout);
+
+  printf("DEBUG: About to initialize THD...\n");
+  fflush(stdout);
+
+
+  if (init_early_variables() ||
+     init_thread_environment() ||
+     mysql_init_variables() ||
+     init_character_sets() ||
+     init_plugin_system_complete())  // This is the key addition
```

Comment:
only mysql_init_variables can really fail. I think it's reasonable to convert the rest of the functions to void and move outside if